File: notes/2021/10/19/BigFraction.java
import java.math.BigInteger;

/**
 * Class used to represent a fraction using two BigIntegers.
 */
public class BigFraction
{
    /**
     * BigFraction with a value of 0/1.
     */
    public final static BigFraction ZERO = new BigFraction();
    /**
     * BigFraction with a value of 1/1.
     */
    public final static BigFraction ONE = BigFraction.valueOf(1, 1);
    /**
     * BigFraction with a value of 1/2.
     */
    public final static BigFraction HALF = BigFraction.valueOf(1, 2);

    private final BigInteger num;
    private final BigInteger denum;

    /**
     * Create a BigFraction from two BigIntegers. Fractions will be simplified based on GCD.
     * @param num The numerator to use for the fraction.
     * @param denum The denominator for the fraction.
     * @throws IllegalArgumentException When a denominator of zero is provided.
     */
    public BigFraction(BigInteger num, BigInteger denum)
    {

        if (denum.compareTo(BigInteger.ZERO) == 0) {
            throw new IllegalArgumentException("Denominator cannot be zero.");
        }

        if (denum.compareTo(BigInteger.ZERO) < 0) {
            denum = denum.abs();
            num = num.multiply(BigInteger.valueOf(-1));
        }

        BigInteger divisor = denum.gcd(num);
        if (divisor.compareTo(BigInteger.ONE) != 0) {
            num = num.divide(divisor);
            denum = denum.divide(divisor);
        }

        this.num = num;
        this.denum = denum;

    }

    /**
     * Creates a BigFraction with the value of 0/1.
     */
    public BigFraction()
    {
        this(BigInteger.ZERO, BigInteger.ONE);
    }

    /**
     * Creates a BigFraction from two longs.
     * @param n The numerator for the fraction.
     * @param d The denominator for the fraction.
     * @return a BigFraction with the specified numerator/denominator.
     */
    public static BigFraction valueOf(long n, long d) {
        return new BigFraction(BigInteger.valueOf(n), BigInteger.valueOf(d));
    }
    
    /**
     * Adds two BigFractions together.
     * @param that The other BigFraction to add.
     * @return A new BigFraction containing the sum of the two BigFractions.
     */
    public BigFraction add(BigFraction that)
    {
        BigInteger top = num.multiply(that.denum).add(denum.multiply(that.num));
        BigInteger bottom = denum.multiply(that.denum);
        return new BigFraction(top, bottom);
    }

    /**
     * Subtract a BigFraction from this BigFraction.
     * @param that The BigFraction that will subtract.
     * @return A new BigFraction with (Current - that) as a value.
     */
    public BigFraction subtract(BigFraction that)
    {
        BigInteger top = num.multiply(that.denum).subtract(denum.multiply(that.num));
        BigInteger bottom = denum.multiply(that.denum);
        return new BigFraction(top, bottom);
    }

    /**
     * Multiply two BigFractions together.
     * @param that The BigFraction to multiply by.
     * @return The BigFraction with the product of that and the current fraction.
     */
    public BigFraction multiply(BigFraction that)
    {
        return new BigFraction(num.multiply(that.num), denum.multiply(that.denum));
    }

    /**
     * Divide the current BigFraction by that
     * @param that The divisor fraction.
     * @return The BigFraction with the value of (this / that).
     */
    public BigFraction divide(BigFraction that)
    {
        return new BigFraction(num.multiply(that.denum), denum.multiply(that.num));
    }

    /**
     * Check if two BigFractions equal each other in value. 
     * Will return false if obj is not a BigFraction.
     */
    @Override
    public boolean equals(Object obj)
    {
        if(obj == this)
        {
            return true;
        }
        if (!(obj instanceof BigFraction)) {
            return false;
        }
        BigFraction other = (BigFraction) obj;
        return (this.num.equals(other.num) && this.denum.equals(other.denum));

    }

    /**
     * Returns the BigFraction in forms of <code>numerator/denominator</code>.
     */
    @Override
    public String toString()
    {
        return String.format("%s/%s", this.num, this.denum);
    }

    // public static void main(String[] args)
    // {
    //     System.out.println("Fraction");
    //     BigFraction boo = BigFraction.valueOf(3, 4);
    //     System.out.println(boo);

    //     System.out.println("Negative Denominator");
    //     boo = BigFraction.valueOf(3, -4);
    //     System.out.println(boo);

    //     System.out.println("Simplification from 1000/2000");
    //     boo = BigFraction.valueOf(1000, 2000);
    //     System.out.println(boo);

    //     System.out.println("Addition");
    //     BigFraction x = new BigFraction();
    //     BigFraction y = BigFraction.valueOf(3, 2);
    //     System.out.printf("%s + %s = %s\n", x, y, x.add(y));
        
    //     System.out.println("Multiplication");
    //     x = BigFraction.valueOf(3, 4);
    //     y = BigFraction.valueOf(1, 2);
    //     System.out.printf("%s * %s = %s\n", x, y, x.multiply(y));
        
    //     System.out.println("Division");
    //     y = BigFraction.valueOf(2, 1);
    //     System.out.printf("%s / %s = %s", x, y, x.divide(y));
    // }
}

File: notes/2021/10/19/BigFraction.py
import math

class BigFraction:
    def __init__(self, num=0, denum=1):
        
        if denum == 0:
            raise ValueError("Denominator cannot be equal to 0.")
        
        if (gcd := math.gcd(denum, num)) != 1:
            num = num//gcd
            denum = denum//gcd
        
        if denum < 0:
            denum *= -1
            num *= -1
        
        
        self.num = num
        self.denum = denum

    def __str__(self):
        return f'{self.num}/{self.denum}'
    
    def __eq__(self, o):
        if not isinstance(o, BigFraction):
            return False
        if o is self:
            return True
        return self.num == o.num and self.denum == o.denum

    def __add__(self, o):
        if not isinstance(o, BigFraction):
            raise ArithmeticError("Cannot add BigFraction and "+o.__class__.__name__)
        return BigFraction(self.num*o.denum+self.denum*o.num, self.denum*o.denum)
    

if __name__ == '__main__':
    # default constructor
    x = BigFraction()
    print(x)
    assert str(x) == "0/1"
    
    # negative swap
    x = BigFraction(3, -4)
    print(x)
    assert str(x) == "-3/4"
    
    # two negatives
    x = BigFraction(-4, -3)
    print(x)
    assert str(x) == "4/3"
    
    # simplify
    x = BigFraction(100, -200)
    print(x)
    assert str(x) == "-1/2"
    
    # zero denum
    try:
        x = BigFraction(1, 0)
    except ValueError as e:
        assert str(e) == "Denominator cannot be equal to 0."
        
    # equals
    x = BigFraction(1, 2)
    y = BigFraction(2, 4)
    
    print(f'{x} == {y}')
    assert x == y
    
    y = BigFraction(2, 3)
    print(f'{x} != {y}')
    assert x != y
    
    # add
    x = BigFraction(1, 2) + BigFraction(1, 2)
    print(f'1/2 + 1/2 = {x}')
    assert str(x) == "1/1"
